### DISCUSSION TITLE
Can I automatically generate static website on my webspace?

### DISCUSSION THREAD

**Discussion by CryssDOTnet** [datenstrom/community#139] [Ask a question] [answered]
Have some domain like admin.mydomain.de with my yellow installation where I maintain the system. Then I want mydomain.de to present only static html to the website visitor – the output of "php yellow.php build".

This works for me but I had to install yellow on my localhost and upload the static output.
Is there no way to make static output from an installation on my webspace? I don't have access to a shell for the php command from my webspace provider.

The goal is to make the site even faster and a bit more secure by omitting php. Only pure and clear  html.

Chris

**Comment by richi** [comment]
I love it! Imagine under dynamic.domain.net you can access the dynamic (editable) version of your site and with a command on a kind of "cli" page, yellow would bake the static site for static.domain.net.

Great idea IMO.

richi

**Comment by CryssDOTnet** [comment]
richi said it all. This is exactly what I want.

**Comment by CryssDOTnet** [comment]
As I said before, I don't have shell access and I am not a developer. I go for the current solution of installing a local server.

Happy new year!

**Comment by markseuffert** [comment]
There are several ways to deploy a website to your web hosting provider:
1. FTP or SFTP, the boring manual way... or in the future via a plugin
2. [cronjob](https://help.ubuntu.com/community/CronHowto), to make a static build in regular intervals, requires shell access
3. [rsync](http://www.tecmint.com/sync-two-apache-websites-using-rsync/), to sync your files at the command line, requires shell access
4. Dropbox to sync your files automatically, requires a web hosting that offers it

I am afraid to say that most good ways require shell access. If you have shell access then there are many possibilities. A cronjob is a good way, if you want to edit your website in one subdomain and publish it later to the official domain. Syncing files with rsync/Dropbox is a handy way, if you want to edit your website on one computer and publish it later to the official server. 

Hope this helps.

**Comment by richi** [comment]
Actually there is no need for ftp, dropbox or such, since both domains are hosted on the same machine, only in different directories. What is needed is just the ability to start the baking without shell access. Converting the cli plugin to a html plugin so to speak. Since my PHP knowledge is limited and rusty, I would at least need an educated guess wether this is possible at all. 
What do you think?

**Comment by CryssDOTnet** [comment]
What about adding a auto-generate "script" (enabled/disabled in the config) to automatically output a static version on visiting the site (in yellow.php)

I will try to add the php build command in yellow.php later.

**Comment by richi** [comment, marked as answer]
Here we go:

https://github.com/richi/yellowcms-cli

This is a small and experimental plugin for utilising the command line interface without shell access.
This is the first shoot. It might eat all your date or nuke your server. Use at your own risk.

Cheers, richi

**Comment by richi** [comment]
If your site is running apache you can use mod_rewrite to serve files directly from the cache directory. Use the cli plugin https://github.com/richi/yellowcms-cli and run a plain "build" command. Yellow will fill the cache directory. Using the original .htaccess file yellow.php will check the cache first, but is still getting started and does the serving. With a modified .htaccess file any cached file will directly be served through apache without bothering yellow.php at all. On a shared host this might speed up things. Here is an example .htaccess file:

```
<IfModule mod_rewrite.c>
RewriteEngine on

DirectoryIndex index.html yellow.php

# check the cache and serve if the file exists
RewriteCond %{DOCUMENT_ROOT}/cache%{REQUEST_URI}/index.html -f [NC] 
RewriteRule ^ cache%{REQUEST_URI}/index.html [L]

RewriteRule ^(cache|content|system)/ error [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^ yellow.php [L]
</IfModule>
```

Cheers, richi

**Comment by CryssDOTnet** [comment]
Thanks! I will have a look into that later. :-)

**Comment by wunderfeyd** [comment]
@CryssDOTnet: Have you already tested richis plugin? Does it work for you? Just want to know.

**Comment by textplusdata** [comment]
Hi,
I tested on a local installation on windows. I deleted my path access for the Windows command line tool before testing so that should have blocked shell access.

Build and Clean both appear to work on a vanilla installation (no testing with blog plugin).

One issue - if you enter the name of a directory to clean that doesn't exist, you still get a message that it has been cleaned (will put this in the plugin issue queue).

Thanks! This plugin is really good for windows users (particularly working with portable drives) since then we don't have to muck around changing paths in the windows system config which is pretty nasty.

**Comment by richi** [comment]
I fixed a problem, that caused a lot of garbage files: https://github.com/richi/yellowcms-cli

And this is a better .htaccess file for serving  the data directly out of the cache.

```
<IfModule mod_rewrite.c>

RewriteEngine on

DefaultType text/html
DirectoryIndex index.html yellow.php

RewriteCond %{DOCUMENT_ROOT}/cache%{REQUEST_URI}index.html -f
RewriteRule ^ /cache%{REQUEST_URI}index.html [L] 
RewriteCond %{DOCUMENT_ROOT}/cache%{REQUEST_URI} -f
RewriteRule ^ /cache%{REQUEST_URI} [L] 

RewriteRule ^(content|system)/ error [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^ yellow.php [L]

</IfModule>
```

On some shared servers (STRATO for example) DOCUMENT_ROOT is not working. If this is the case simply replace `%{DOCUMENT_ROOT}` with the real path.
